Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/debuglog/DebugLogger.kt

```diff
@@ -0,0 +1,161 @@
+package de.rki.coronawarnapp.bugreporting.debuglog
+
+import android.annotation.SuppressLint
+import android.app.Application
+import android.content.Context
+import android.util.Log
+import de.rki.coronawarnapp.util.di.ApplicationComponent
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.yield
+import timber.log.Timber
+import java.io.File
+
+@SuppressLint("LogNotTimber")
+@Suppress("BlockingMethodInNonBlockingContext")
+object DebugLogger : DebugLoggerBase() {
+
+    private val scope = DebugLoggerScope
+    private lateinit var context: Context
+
+    private val debugDir by lazy {
+        File(context.cacheDir, "debuglog").also {
+            if (!it.exists()) it.mkdir()
+        }
+    }
+
+    private val triggerFile by lazy { File(debugDir, "debug.trigger") }
+
+    internal val runningLog by lazy { File(debugDir, "debug.log") }
+    val sharedDirectory by lazy { File(debugDir, "shared") }
+
+    private val mutex = Mutex()
+
+    private var logJob: Job? = null
+    private var logTree: DebugLogTree? = null
+    private var isDaggerReady = false
+
+    fun init(application: Application) {
+        context = application
+
+        if (triggerFile.exists()) {
+            Timber.tag(TAG).i("Trigger file exists, starting debug log.")
+            runBlocking { start() }
+        }
+    }
+
+    /**
+     * To censor unique data, we need to actually know what to censor.
+     * So we buffer log statements until Dagger is ready
+     */
+    fun setInjectionIsReady(component: ApplicationComponent) {
+        Timber.tag(TAG).i("setInjectionIsReady()")
+        component.inject(this)
+        isDaggerReady = true
+    }
+
+    val isLogging: Boolean
+        get() = logJob?.isActive == true
+
+    suspend fun start(): Unit = mutex.withLock {
+        Timber.tag(TAG).d("start()")
+
+        if (isLogging) {
+            Timber.tag(TAG).w("Ignoring start(), already running.")
+            return@withLock
+        }
+
+        logJob?.cancel()
+        logTree?.let { Timber.uproot(it) }
+
+        DebugLogTree().apply {
+            Timber.plant(this)
+            logTree = this
+
+            if (!runningLog.exists()) {
+                runningLog.parentFile?.mkdirs()
+                if (runningLog.createNewFile()) {
+                    Timber.tag(TAG).i("Log file didn't exist and was created.")
```

Comment:
Would be cool if you were to log the 
- app version
- ENF version
- device details, such as Android version, device type, manufacturer, ... 

at the beginning of each log.

Reply:
Device infos are already logged when the app process is started.
I've changed it to also log them again when the debug log is started manually, otherwise they would only be visible after the next app start.
I've also added the ENF version there, which we do not print on app process start due to being more expensive and prone to side effects.